src/core/xccl_lock_free_tasks_queue.c

```diff
@@ -0,0 +1,153 @@
+#include "xccl_lock_free_tasks_queue.h"
+
+xccl_status_t lf_tasks_queue_init(xccl_progress_queue_t *handle) {
+    handle->ctx = (void *) malloc(sizeof(xccl_lf_tasks_queue_t));
+    xccl_lf_tasks_queue_t *ctx = (xccl_lf_tasks_queue_t *) handle->ctx;
+
+    ctx->deque_size = (xccl_lib_global_config.lf_tasks_queue_size / LINE_SIZE) + 1;
+
+    ctx->tasks = (ucc_coll_task_t ***) calloc(2 * ctx->deque_size, sizeof(ucc_coll_task_t **));
+    if (ctx->tasks == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+    ctx->tasks[0] = (ucc_coll_task_t **) calloc(LINE_SIZE, sizeof(ucc_coll_task_t *));
+    if (ctx->tasks[0] == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+    ctx->tasks[ctx->deque_size] = (ucc_coll_task_t **) calloc(LINE_SIZE, sizeof(ucc_coll_task_t *));
+    if (ctx->tasks[ctx->deque_size] == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+
+    ucs_spinlock_init(&(ctx->increase_queue_locks[0]), 0);
+    ucs_spinlock_init(&(ctx->increase_queue_locks[1]), 0);
+    ucs_spinlock_init(&(ctx->locked_queue_lock), 0);
+    ucs_list_head_init(&ctx->locked_queue);
+    ctx->which_pool = 0;
+
+    handle->api.progress_queue_enqueue        = &lf_tasks_queue_insert;
+    handle->api.progress_queue_progress_tasks = &lf_tasks_queue_progress;
+    handle->api.progress_queue_cleanup        = &lf_tasks_queue_cleanup;
+    return XCCL_OK;
+}
+
+xccl_status_t increase_tasks(xccl_lf_tasks_queue_t *ctx, int i) {
+    ucs_spin_lock(&ctx->increase_queue_locks[i / ctx->deque_size]);
+    if (ctx->tasks[i] == 0) {
+        ctx->tasks[i] = (ucc_coll_task_t **) calloc(LINE_SIZE, sizeof(ucc_coll_task_t *));
+        if (ctx->tasks[i] == NULL) {
+            return XCCL_ERR_NO_MEMORY;
+        }
+    }
+    ucs_spin_unlock(&ctx->increase_queue_locks[i / ctx->deque_size]);
+    return XCCL_OK;
+}
```

Comment:
@vspetrov @Sergei-Lebedev Regarding the issue we spoke about with Manju, should we keep the increase size feature, and grow the list slow, or should we just initialize with MAX_SIZE and that it.
I apologize, because we only spoke about memory waste aspects, which aren’t so bad, but I forgot to bring up some other concerning issue – because we are iterating over the list (many times until it’s end, when one of the lists are empty) – there is a great value of keeping the list small as possible, in order to save time.
the increase list ability handles this issue well, because if LINE_SIZE = 8 for example, and MAX_SIZE is 128, most of the time actual size will be 8 or 16, and not 128. If we will init to 128 from the very beginning, we will need to iterate all in every pop operation.
Because of that, I kept the increase size design for now. If you want to discuss again, or decide elsewise, NP.